sql: TestTrace failed

sql.TestTrace [failed](https://mesolite.cluster.engflow.com/invocations/default/7561edb2-bbfc-4436-a96b-bc7ca79c5e71?testReportRun=4&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [3dc59b8f9bbac2f0b97eb530eba8bd588d3bc289](https://github.com/cockroachdb/cockroach/commits/3dc59b8f9bbac2f0b97eb530eba8bd588d3bc289):


```
        executing Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 8e05449d], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can't serve follower read; closed timestamp too low by: 3.512907544s; maxClosed: 1778127439.266112182,0 ts: 1778127442.279019726,0 uncertaintyLimit: 1778127442.779019726,0
        replica.Send got error: [NotLeaseHolderError] lease held by different store; r103: replica (n3,s3):3 not lease holder; current lease is repl=(n2,s2):2 seq=2 start=1778127438.365253694,0 type=LeaseLeader term=6 min-exp=1778127448.128126065,0 pro=1778127442.188269872,0 acq=Request
        proxy request for Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 8e05449d], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760] after local error [NotLeaseHolderError] lease held by different store; r103: replica (n3,s3):3 not lease holder; current lease is repl=(n2,s2):2 seq=2 start=1778127438.365253694,0 type=LeaseLeader term=6 min-exp=1778127448.128126065,0 pro=1778127442.188269872,0 acq=Request: "sql txn" meta={id=8e05449d key=/Min iso=Serializable pri=0.05007302 epo=0 ts=1778127442.279019726,0 min=1778127442.279019726,0 seq=0} lock=false stat=PENDING rts=1778127442.279019726,0 gul=1778127442.779019726,0 obs={n3@1778127442.282500373,0}
        === operation:dist sender send gid:342862 _verbose:1 node:3 tenant:10
        processing a proxy request to repl=(n2,s2):2 seq=2 start=1778127438.365253694,0 type=LeaseLeader term=6 min-exp=1778127448.128126065,0 pro=1778127442.188269872,0 acq=Request
        clearing entries overlapping r103:/Tenant/10/Table/{48-107/1/5} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=17]
        sending request to 127.0.0.1:45635
        sending batch request
        === operation:/cockroach.roachpb.KVBatch/Batch gid:344513 _verbose:1 node:2 span.kind:server request:Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 8e05449d], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760]
        node received request: 1 Scan
        executing Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 8e05449d], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760]
        read-only path
        sequencing request
        acquiring latches
        scanning lock table for conflicting locks
        can drop latches early for batch (Scan [/Tenant/10/Table/106/1,/Tenant/10/Table/106/2), [txn: 8e05449d], [can-forward-ts], [max_span_request_keys: 0], [target_bytes: 10485760]); scanning lock table first to detect conflicts
        lock table scan stats: seeked 1 times (1 internal); stepped 0 times (0 internal)
        lock table scan complete without conflicts; dropping latches early
        executing read-only batch
        evaluated Scan command header:<key:"\376\222\362\211" end_key:"\376\222\362\212" > scan_format:BATCH_RESPONSE , txn="sql txn" meta={id=8e05449d key=/Min iso=Serializable pri=0.05007302 epo=0 ts=1778127442.279019726,0 min=1778127442.279019726,0 seq=0} lock=false stat=PENDING rts=1778127442.279019726,0 gul=1778127442.779019726,0 obs={n2@1778127442.282824466,0} : resp=header:<> col_batches:<> , err=<nil>
        read completed
        n2 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
        node sending response
        received batch response
        proxy result is (err: <nil>), *kvpb.ScanResponse <nil>
        proxy request completed
        node sending response
        received batch response
        r_u:0.625 
        recording span to refresh: /Tenant/10/Table/106/{1-2}
        ComponentStats{ID: {5b6fab3b-88bd-4d53-a217-f64ddeeb5e0e PROCESSOR 0 3 }, KV nodes: n2, KV regions: test, KV time: 1ms, KV rows decoded: 0, KV bytes read: 0 B, KV gRPC calls: 1, KV pairs read: 0, MVCC step count (ext/int): 0/0, MVCC seek count (ext/int): 1/1, rows output: 0}
        ComponentStats{ID: {5b6fab3b-88bd-4d53-a217-f64ddeeb5e0e FLOW 0 3 test}}
        execution ends
        rows affected: 0
        === operation:commit sql txn gid:344942 _verbose:1 node:sql3 client:127.0.0.1:50578 hostssl: user:root
        AutoCommit. err: <nil>
        in instrumenationHelper.Finish
        releasing 4 descriptors
        release: 100("defaultdb",010180c809a2ac09944f5bbe2781c24e6ffc9d) ver=1:0,0, refcount=0
        release: 104("test",010180c809a2ac09944f5bbe2781c24e6ffc9d) ver=1:0,0, refcount=0
        release: 105("public",010180c809a2ac09944f5bbe2781c24e6ffc9d) ver=1:0,0, refcount=0
        release: 106("foo",010180c809a2ac09944f5bbe2781c24e6ffc9d) ver=1:0,0, refcount=0
        [NoTxn pos:10] executing ExecStmt: SET TRACING = off
        executing: SET TRACING = off in state: NoTxn
--- FAIL: TestTrace/Session/TracingOff/node-2 (0.04s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTrace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63691

--- a/pkg/sql/trace_test.go
+++ b/pkg/sql/trace_test.go
@@ -214,6 +214,15 @@ func TestTrace(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	// Wait for each node's statement_hints rangefeed to complete its initial
+	// scan. Otherwise, the first query on a node will fall through to a DB read
+	// in MaybeGetStatementHints (because hintedHashes is still nil) and pollute
+	// the trace with internal-query spans like colbatchscan / colindexjoin.
+	for i := 0; i < numNodes; i++ {
+		hc := cluster.ApplicationLayer(i).ExecutorConfig().(sql.ExecutorConfig).StatementHintsCache
+		hc.Await(context.Background())
+	}
+
 	for _, test := range testData {
 		optionalSpans := append([]string{}, alwaysOptionalSpans...)
 		// Check if DRPC is enabled. When DRPC is enabled, remote KV requests

